### PR TITLE
feat: add SSR deployment adapters for Cloudflare Workers, Vercel, and Deno

### PR DESCRIPTION
**Pull request by advl**
## Done

Three **private** packages bridging `@canonical/react-ssr` renderers to platform deployment targets (no npm publication).

> **These packages will likely be relocated outside the main `pragma` monorepo** once the SSR deployment story settles. They live here for now to iterate alongside `@canonical/react-ssr`; treat their location as provisional. (Same note is in each adapter's README.)

**Adapters**
- **`@canonical/ssr-adapter-deno`** — `createHandler(config)` for `Deno.serve()`.
- **`@canonical/ssr-adapter-cloudflare`** — `createFetchHandler(config)` for Workers `fetch(request, env, ctx)`.
- **`@canonical/ssr-adapter-vercel`** — `createNodeHandler` (serverless) and `createEdgeHandler` (edge).

Each is a pure SSR router over an ordered `routes[]` list (first match wins). The client `AbortSignal` is forwarded to the renderer on every handler so a disconnect cancels the render. Each package is **100% covered** (lines/branches/functions, enforced in vitest).

**Static assets are delegated to the platform** (researched against official docs — the approach is deliberately asymmetric because the platforms differ):
- **Cloudflare** and **Vercel** serve static files at the edge *before* the SSR function runs, so those adapters take no static-asset config — Cloudflare via Workers Static Assets (`[assets]` in `wrangler.toml`), Vercel via its CDN (`.vercel/output/static`).
- **Deno Deploy** has no edge static layer for a dynamic app, so the Deno adapter serves files in-process — streaming via the native `Deno.open().readable`.

**Multiple backends** (e.g. a sitemap alongside the app) are served from one handler by listing several `routes[]` entries (sitemap before the `/*` catch-all) — the required model on Deno, preferred on Cloudflare, acceptable on Vercel.

**Adversarial, docs-grounded review fixes** (each with a regression test):
- Vercel Node: forward a client-disconnect `AbortSignal` to the renderer/`Request`; stream the response via `stream.pipeline(Readable.fromWeb(...), res)` for backpressure + error/abort handling.
- Deno: stream assets via `Deno.open().readable` (node:fs fallback under tests).
- The earlier R2 key-prefix / metadata / GET-gating fixes were retired along with the Cloudflare R2 path when static serving moved to the edge.

**Docs:** a README per adapter, each documenting its static-asset approach and cross-referencing the siblings on why they differ, plus a multiple-backends section. Adds the missing `@types/node` to the Deno package.

Fixes the stale base of #597 (rebased onto current `main`; the shared adapter types from #596 are already merged).

## QA

For each adapter package, from the repo root or the package directory:
- `bun run check` — passes (biome + tsc).
- `bun run test:coverage` — 100% (lines/branches/functions).
- `bun run build` — clean tsc build.

Full CI mirror run locally against `origin/main`:
- `bunx nx affected -t check` — 52 projects, pass.
- `bunx nx affected -t build:all` — 42 projects, pass.
- `bunx nx affected -t test` — 36 projects, pass.

Behavioural checks covered by the new regression tests: route ordering / first-match, renderer `statusCode` propagation, `AbortSignal` forwarding + abort-on-disconnect (Vercel Node, Deno), Node response streaming/backpressure, Deno native-`Deno.open` path, and the Cloudflare GET-only edge cache.

### PR readiness check

- [x] PR should have one of the following labels: `Feature 🎁`.
- [x] PR title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
- [x] The code follows the appropriate [code standards](https://github.com/canonical/code-standards).
- [x] All packages define the required scripts in `package.json`:
  - [x] All packages: `check`, `check:fix`, and `test`.
  - [x] Packages with build steps: `build` and `build:all`.
- [x] These packages are **private** (`"private": true`) — no first-time publish needed.
- [x] No visual change — `no visual change` label added to skip Chromatic.

## Screenshots

N/A — no visual change (CLI/SSR runtime packages).

### CODE REVIEW

File: packages/react/ssr-adapter-cloudflare/src/lib/handler/createFetchHandler.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    // cache.put rejects on non-GET requests, so the edge cache is GET-only.
+    const cacheable = useCache && request.method === "GET";
+
+    // 1. Static assets from R2
+    for (const asset of config.staticAssets ?? []) {
+      if (url.pathname.startsWith(asset.urlPrefix)) {
+        const tail = url.pathname
+          .slice(asset.urlPrefix.length)
+          .replace(/^\//, "");
+        if (!tail) continue;
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/react/ssr-adapter-cloudflare/src/lib/handler/createFetchHandler.ts]: Claude code: Resolved by the static-delegation refactor — the Cloudflare adapter no longer serves static assets at all (Workers Static Assets handles them at the edge), so this prefix-matching code was removed. No `startsWith` asset lookup remains in `createFetchHandler.ts`.

File: packages/react/ssr-adapter-deno/src/lib/handler/createHandler.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * ```
+ */
+export function createHandler(config: DenoAdapterConfig) {
+  return async (request: Request): Promise<Response> => {
+    const url = new URL(request.url);
+
+    for (const asset of config.staticAssets ?? []) {
+      if (url.pathname.startsWith(asset.urlPrefix)) {
+        const relativePath = url.pathname.slice(asset.urlPrefix.length);
+        const filePath = `${asset.directory}${relativePath}`;
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/react/ssr-adapter-deno/src/lib/handler/createHandler.ts]: Claude code: Good catch — fixed in 2554277a. The Deno adapter now matches on a path-segment boundary: it normalises a trailing slash on `urlPrefix` and only matches when `pathname === prefix` or the remainder starts with `/`, so `/assets` no longer matches `/assets2/...` (which also fixes the broken join). Added a regression test asserting `/assets2/main.js` falls through to the renderer.

File: packages/react/ssr-adapter-cloudflare/src/lib/handler/createFetchHandler.tests.ts [hunk 3]
```diff
@@ -0,0 +1,446 @@
+import { describe, expect, it, vi } from "vitest";
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/react/ssr-adapter-cloudflare/src/lib/handler/createFetchHandler.tests.ts]: Claude code: Fixed in 2554277a. `beforeEach` is now part of the top-level vitest import and the stray bottom-of-file import was removed.

File: packages/react/ssr-adapter-deno/src/lib/handler/createHandler.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        try {
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/react/ssr-adapter-deno/src/lib/handler/createHandler.ts:53]: Claude code: Both points addressed. The path-segment boundary was fixed in 2554277a (`/assets` no longer matches `/assets2/...`). Path traversal is now guarded in 9cf9a976: the asset path is decoded and any `..` segment (literal or percent-encoded, e.g. `..%2f`, `%2e%2e`) is rejected before opening the file, with malformed encoding bailing too. Regression tests cover the boundary case and the traversal/encoded/malformed cases.

File: packages/react/ssr-adapter-vercel/src/lib/handler/createNodeHandler.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  for (const [key, value] of Object.entries(req.headers)) {
+    if (value != null) {
+      headers.set(key, Array.isArray(value) ? value.join(", ") : value);
+    }
+  }
+  return new Request(url.toString(), {
+    method: req.method ?? "GET",
+    headers,
+    signal,
+  });
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/react/ssr-adapter-vercel/src/lib/handler/createNodeHandler.ts]: Claude code: Fixed in 9cf9a976. The Node handler now forwards the request body for non-GET/HEAD methods by streaming the `IncomingMessage` via `Readable.toWeb(req)` with `duplex: "half"`, so form posts and API routes reach renderers with their body — matching the Cloudflare/Deno/Edge handlers. GET/HEAD still carry no body (asserted by a regression test alongside the non-GET body-forwarding test).
